File: Data Structures & Algorithms/depth-of-binary-tree/submission-1.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

public class Solution {
    public int MaxDepth(TreeNode root) {
        return dfs(root);
    }

    public int dfs(TreeNode node) {
        if (node is null) {
            return 0;
        }

        int left = 0;
        int right = 0;
        if (node.left is not null) {
            left = dfs(node.left);
        }

        if (node.right is not null) {
            right = dfs(node.right);
        }

        return 1 + Math.Max(left, right);
    }
}

File: Data Structures & Algorithms/binary-tree-diameter/submission-1.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        def dfs(node):
            if not node:
                return 0, 0
            
            left_height, left_diameter = dfs(node.left)
            right_height, right_diamater = dfs(node.right)

            height = 1 + max(left_height, right_height)

            through_root = left_height + right_height
            diamater = max(left_diameter, right_diamater, through_root)
            return height, diamater
        
        return dfs(root)[1]